okhttp/src/main/java/io/grpc/okhttp/OkHttpClientTransport.java

```diff
@@ -721,8 +732,14 @@ void finishStream(
           frameWriter.rstStream(streamId, ErrorCode.CANCEL);
         }
         if (status != null) {
-          stream.transportState().transportReportStatus(status, stopDelivery,
-              trailers != null ? trailers : new Metadata());
+          boolean isCancelled =
+              (status.getCode() == Code.CANCELLED || status.getCode() == Code.DEADLINE_EXCEEDED);
+          stream
+              .transportState()
+              .transportReportStatus(
+                  status,
+                  isCancelled || stopDelivery,
```

Comment:
This is complicated and surprising. I had hoped only the RST stream method I mentioned earlier would have this logic.